Add support for durable functions

ChristianEder opened the pull request:
This PR adds support for creating Azure Durable Functions more easily - see example added.

I remember that there was a 'examples_test.go' file where I'd like to add my example to - but since my last PR this changed quite a bit and I'm not sure what is required there.

mikhailshilkov commented:
@ChristianEder Could you please rebase on master (it's v2.0 now) and add a test that would be a copy of https://github.com/pulumi/pulumi-azure/blob/master/examples/examples_test.go#L84 but for your new example?

ChristianEder commented:
@mikhailshilkov done.

ChristianEder commented:
hmm the build seems to fail, but I can't really figure out why - the only error it states is that there are changes in the working tree - in the files that existed before my PR, which i changed and another one that got somehow added by the build:

- changed: `sdk/nodejs/appservice/index.ts`
- added by build: `packages-microsoft-prod.deb`

any ideas?

mikhailshilkov commented:
@ChristianEder Yes, you need to add your new file to the list here: https://github.com/pulumi/pulumi-azure/blob/master/resources.go#L1416

ChristianEder commented:
@mikhailshilkov hmm i added the new mixin file to the `resources.go` file, but the build still fails, now with changes in the ` sdk/nodejs/tsconfig.json` file (which I didnt change in my PR), and still the `packages-microsoft-prod.deb` file.

mikhailshilkov commented:
Yeah, now you need to change `tsconfig.json` and commit it. It would be done by the tooling if you could run `make` but I guess you are on Windows and then it's non-trivial.

ChristianEder commented:
wohoo, green! Thanks @mikhailshilkov for your help, I totally forgot about all these places that I need to add new mixins to...

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+        const trigger = {
+            name: args.activityInputName,
+            type: "activityTrigger",
+            direction: "in"
```
```suggestion
            direction: "in",
```

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+}
+
+export interface DurableEntityFunctionContext extends mod.Context<mod.Result> {
+    df: any
```
```suggestion
    df: any;
```

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+export class DurableEntityFunction extends mod.Function<DurableEntityFunctionContext, void, void> {
+    constructor(name: string, args: DurableEntityFunctionArgs) {
+
+        if(args.callback){
```
```suggestion
        if (args.callback) {
```

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+        const trigger = {
+            name: "context",
+            type: "entityTrigger",
+            direction: "in"
```
```suggestion
            direction: "in",
```

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+        this.binding = {
+            name,
+            type: "durableClient",
+            direction: "in"
```
```suggestion
            direction: "in",
```

mikhailshilkov commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+        this.binding = {
+            name,
+            type: "orchestrationClient",
+            direction: "in"
```
```suggestion
            direction: "in",
```

mikhailshilkov commented on examples/durable-functions/index.ts:
```diff
@@ -0,0 +1,69 @@
+        }),
+        new azure.appservice.HttpFunction("viewCount", {
+            route: "views",
+            callback: async (con, req) => {
```
Let's call it `context` everywhere

mikhailshilkov commented on examples/durable-functions/index.ts:
```diff
@@ -0,0 +1,69 @@
+    ]
+});
+
+export const url = app.endpoint;
```
Do you mind adding a short instruction on how to test the deployed functions?

ChristianEder replied:
added a comment explaining how to call the durable clients via http

ChristianEder commented on examples/durable-functions/index.ts:
```diff
@@ -0,0 +1,69 @@
+import * as azure from "@pulumi/azure";
+import * as df from 'durable-functions';
+
+const resourceGroup = new azure.core.ResourceGroup("chedDurable", {
```
oh my. I'll rename

ChristianEder commented on examples/durable-functions/index.ts:
```diff
@@ -0,0 +1,69 @@
+        new azure.appservice.DurableActivityFunction("SayHello", {
+            activityInputName: "name",
+            callback: (context) => {
+                context.done(null, "Hello, " + context.bindings.name);
```
should be, will change

ChristianEder commented on examples/examples_test.go:
```diff
@@ -26,6 +26,16 @@ func TestAccMinimal(t *testing.T) {
+	test := getJSBaseOptions(t).
+		With(integration.ProgramTestOptions{
+			Dir: path.Join(getCwd(t), "durable-functions"),
+			//RunUpdateTest: true, // temporarily disabled as upgrade tests are not able to run until we release 2.0.0
```
I will change this throughout the file

ChristianEder commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+import * as mod from ".";
+
+export interface DurableOrchestratorFunctionContext extends mod.Context<mod.Result> {
+    df: any
```
Probably not really - unless I not either add a dependency to the 'durable-functions' package (I would assume thats not a solution), or try to replicate the df interface. The package defines the context as

`export interface IOrchestrationFunctionContext extends Context {
    df: DurableOrchestrationContext;
}`

ChristianEder commented on sdk/nodejs/appservice/zMixins_durable.ts:
```diff
@@ -0,0 +1,123 @@
+
+export interface DurableActivityFunctionContext<TActivityInputBinding> extends mod.Context<mod.Result> {
+    bindings: {
+        [key: string]: any;
```
hmm I probably had a reason when I added that but it compiles without it soo... I'll remove it.

ChristianEder commented:
I worked through all your comments, resolved the "easy" ones - do you want to resolve the others / give more feedback?

mikhailshilkov commented on examples/durable-functions/index.ts:
```diff
@@ -1,19 +1,17 @@
             activityInputName: "name",
-            callback: (context) => {
-                context.done(null, "Hello, " + context.bindings.name);
+            callback: async(context) => {
```
```suggestion
            callback: async context => {
```

mikhailshilkov commented on examples/examples_test.go:
```diff
@@ -26,11 +26,21 @@ func TestAccMinimal(t *testing.T) {
+	test := getJSBaseOptions(t).
+		With(integration.ProgramTestOptions{
+			Dir: path.Join(getCwd(t), "durable-functions"),
+			RunUpdateTest: true
```
You have to put `,` at the end of each line (see build errors). Sorry, I didn't realize that 2.0 thing was all over the place, feel free to revert that changes - they shouldn't be your problem.

ChristianEder commented:
> > unless I not either add a dependency to the 'durable-functions' package (I would assume thats not a solution),
> 
> 
> 
> We could add those. We already have dependencies e.g. on `azure-functions-ts-essentials` and `azure-eventgrid`.

From the perspective of a durable function developer, adding the correct type will not add any benefit - because the method containing the durable function itself is wrapped inside a `df.orchestrator`call and gets the correct typing anyway. This type here is only required so that the callback returned by the `df.orchestrator` call gets accepted as a valid callback by the compiler. 

If you‘d prefer having the right types defined anyway, I can add them - but wouldn’t that mean the package being deployed to _any_ function app, even those that don’t use durable functions?
```

mikhailshilkov commented:
Alright, let's keep `any` then. Thanks for the explanation!

ChristianEder commented:
So - build is back to green, all comments resolved. *phew* :-)
